src/proto/policy.proto

```diff
@@ -0,0 +1,172 @@
+// defined in multiple policies and one of the policies is a deny, that action-resource pair is
+// always denied.
+//
+// Next: 3
+message PolicyDocument {
+  // A version number indicating the bowler-proto version this policy was written in. This may
+  // determine how the policy is parsed. Check the bowler-proto releases for information on how
+  // versions may differ. This must be a valid semver version according to Semantic Versioning 2.0
+  // [here](https://semver.org/).
+  string version = 1;
+
+  // The statements that this policy is comprised of.
+  repeated PolicyContainer statements = 2;
+}
+
+// A simple container for the different policy types.
+//
+// Next: 3
+message PolicyContainer {
+  oneof policy {
+    ActionPolicy action_policy = 1;
+    CredentialPolicy credential_policy = 2;
+  }
+}
+
+// Specifies what actions the kernel is allowed to perform and the resources on which those actions
+// may be performed. For example, this policy may specify that the kernel is allowed to flash
+// microcontrollers with a certain vendor ID.
+//
+// Next: 4
+message ActionPolicy {
+  // Whether this policy is an allow (the kernel may perform the actions on the resources) or a deny
+  // (the kernel may not perform the actions on the resources).
+  bool allow = 1;
+
+  // The actions this policy applies to. The action schema is: `<type>:<resource>`. For example,
+  // `flash:uc` is the action for flashing a microcontroller.
+  //
+  // Possible actions are:
+  //
+  // - `flash`: Upload a program to a device (overwriting any program that is already present).
+  //   - `uc`: A microcontroller.
+  repeated string actions = 2;
+
+  // The resources the actions may act on. All the resources specified here must have the same type
+  // as those specified in the actions. There must be exactly one resource specified here for every
+  // resource mentioned in the actions. The resource schema is: `<type>:<scope>:<name>`. For
+  // example, `uc:file:/dev/sda` specifies the microcontroller mounted as the `/dev/sda` file.
+  //
+  // Possible resources are:
+  //
+  // - `uc`: A microcontroller.
+  //   - `name`: The device name. This is the name specified by the device over the Bowler RPC.
+  //     - Schema: `uc:name:<uc name>`
+  //     - E.g. `uc:name:my-uc-name`
+  //   - `file`: An absolute file path in the kernel's local file system.
+  //     - Schema: `uc:file:<file path>`
+  //     - E.g. `uc:file:/dev/sda`
+  //     - E.g. `uc:file:/dev/tty*`
+  //   - `ipv4`: IPv4 address specified as four decimal octets. IP ranges are not accepted. CIDR
+  //   notation is not accepted.
+  //     - Schema: `uc:ipv4:<octet>.<octet>.<octet>.<octet>`
+  //     - E.g. `uc:ipv4:192.168.1.2`
+  //     - E.g. `uc:ipv4:192.168.*`
+  //   - `hid`: Vendor ID and product ID.
+  //     - Schema: `uc:hid:<vid>/<pid>`
+  //     - E.g. `uc:hid:ab12/cd34`
+  repeated string resources = 3;
```

Comment:
Nit: using something structured in protobuf would be better than a custom string serialization format (unless this is widely used in non-protobuf contexts, but again I'd look at it). Largely nitpicking here because I've had API code be nitpicked for this in the past and it did make it better in the end. It also avoids writing (even trivial) parsers and dealing with restrictions on what characters can be used where etc.

Reply:
I have thought about that, but I want to retain the flexibility that text gives the user. I tried making a more structured API and it became slightly messy due to all the ways the user can supply a wildcard.

Reply:
Fair enough